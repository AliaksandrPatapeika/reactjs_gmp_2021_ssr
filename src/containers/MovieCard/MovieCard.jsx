import './MovieCard.less';

import Link from 'next/link';
import PropTypes from 'prop-types';
import React, {useCallback, useState} from 'react';
import {useDispatch} from 'react-redux';

import {showModal, showMovieDetails} from '../../actions/movies';
import Button from '../../components/Button';
import MovieGenres from '../../components/MovieGenres';
import MovieReleaseDate from '../../components/MovieReleaseDate';
import MovieTitle from '../../components/MovieTitle';
import {addDefaultSrc} from '../../utils';
import MovieCardMenu from '../MovieCardMenu';

const MovieCard = ({
  movie,
  movie: {
    id, poster_path, title, release_date, genres
  }
}) => {
  const [showSubMenu, setShowSubMenu] = useState(false);
  const dispatch = useDispatch();

  const showMovieDetailsWindow = useCallback(() => {
    dispatch(showMovieDetails(movie));
  }, [dispatch, movie]);

  const showMovieCardSubMenu = () => {
    setShowSubMenu(true);
  };

  const closeMovieCardSubMenu = () => {
    setShowSubMenu(false);
  };

  const showEditMovieWindow = () => {
    dispatch(showModal('editMovie', movie));
  };

  const showDeleteMovieWindow = () => {
    dispatch(showModal('deleteMovie', movie));
  };

  return (
    <div
      className="movieCardContainer"
      onMouseLeave={closeMovieCardSubMenu}
    >
      <Link
        href={`/film/${id}`}
        // as={`/film/${title}`} // Implement route masking
      >
        <Button className="posterButton" onClick={showMovieDetailsWindow}>
          <img
            src={poster_path || '/img/noImage.png'}
            onError={addDefaultSrc}
            alt="movie card poster"
            className="movieCardPoster"
          />
        </Button>
      </Link>
      <Button className="threeDotsIcon" onClick={showMovieCardSubMenu}>
        <img src="/img/moreButton.svg" alt="movie card menu" />
      </Button>
      <MovieCardMenu
        showSubMenu={showSubMenu}
        closeSubMenu={closeMovieCardSubMenu}
        showEditMovieWindow={showEditMovieWindow}
        showDeleteMovieWindow={showDeleteMovieWindow}
      />
      <div className="movieTitleReleaseDateContainer">
        <MovieTitle title={title} />
        <MovieReleaseDate releaseDate={release_date} />
      </div>
      <MovieGenres genres={genres.join(', ')} />
    </div>
  );
};

MovieCard.propTypes = {
  movie: PropTypes.shape({
    budget: PropTypes.number,
    genres: PropTypes.arrayOf(PropTypes.string),
    id: PropTypes.number,
    overview: PropTypes.string,
    poster_path: PropTypes.string,
    release_date: PropTypes.string,
    revenue: PropTypes.number,
    runtime: PropTypes.number,
    tagline: PropTypes.string,
    title: PropTypes.string,
    vote_average: PropTypes.number,
    vote_count: PropTypes.number
  })
};

MovieCard.defaultProps = {
  movie: null
};

export default MovieCard;
