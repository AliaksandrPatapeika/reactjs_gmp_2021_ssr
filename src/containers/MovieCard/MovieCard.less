@import "../../less/variables";

.movieCardContainer {
	position: relative;
	width: 330px;
	margin-bottom: 40px;
	color: @white;
	font-family: @NetflixSans-Regular;

	.posterButton {
		background-color: transparent;

		.movieCardPoster {
			cursor: pointer;
			max-height: 495px;
			width: 100%;
			object-fit: cover;
		}
	}

	.threeDotsIcon {
		position: absolute;
		right: 10px;
		top: 10px;
		width: 30px;
		height: 30px;
		border: 6px solid @black;
		border-radius: 50%;
		opacity: 0;
		background-color: @black;

		&:focus {
			opacity: 1;
		}
	}

	.movieTitleReleaseDateContainer {
		display: flex;
		justify-content: space-between;
		padding: 8px 0;
	}

	&:hover .threeDotsIcon {
		cursor: pointer;
		opacity: 1;
	}
}
