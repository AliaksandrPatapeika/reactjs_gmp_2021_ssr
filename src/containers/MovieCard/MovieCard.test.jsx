import {head} from 'lodash';
import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';

import MockMovies from '../../tests/mocks/mockData';
import MovieCard from './MovieCard';

describe('MovieCard', () => {
  test('renders correctly', () => {
    const mockStore = configureMockStore();
    const initialState = {};
    const store = mockStore(initialState);
    const movieCard = renderer
      .create(
        <Provider store={store}>
          <MovieCard movie={head(MockMovies)} />
        </Provider>
      )
      .toJSON();

    expect(movieCard)
      .toMatchSnapshot();
  });
});
