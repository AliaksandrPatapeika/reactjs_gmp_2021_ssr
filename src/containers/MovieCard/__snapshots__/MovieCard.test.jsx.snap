// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard renders correctly 1`] = `
<div
  className="movieCardContainer"
  onMouseLeave={[Function]}
>
  <button
    className="posterButton"
    disabled={false}
    id={null}
    onClick={[Function]}
    type="button"
  >
    
    <img
      alt="movie card poster"
      className="movieCardPoster"
      onError={[Function]}
      src="http://localhost:3000/img/01-pulp-fiction.jpg"
    />
  </button>
  <button
    className="threeDotsIcon"
    disabled={false}
    id={null}
    onClick={[Function]}
    type="button"
  >
    
    <img
      alt="movie card menu"
      src="/img/moreButton.svg"
    />
  </button>
  <div
    className="movieTitleReleaseDateContainer"
  >
    <div
      className="movieTitle"
    >
      Pulp Fiction
    </div>
    <div
      className="movieReleaseDate"
      title="Release date: 2004-03-15"
    >
      2004
    </div>
  </div>
  <div
    className="movieGenres"
  >
    Action & Adventure
  </div>
</div>
`;
